File: deepspace/src/deepspace/Hangar.java
package deepspace;
import java.util.ArrayList;

class Hangar {
    private int maxElements;
    private ArrayList<ShieldBooster> shieldBoosters;
    private ArrayList<Weapon> weapons;
    
    
    Hangar(int capacity){
        maxElements=capacity;
        shieldBoosters=new ArrayList<ShieldBooster>();
        weapons=new ArrayList<Weapon>();
    }
    
    Hangar(Hangar h){
        this(h.maxElements);
        
        for(ShieldBooster s: h.shieldBoosters)
            addShieldBooster(s);
        
        for(Weapon w: h.weapons)
            addWeapon(w);
    }
    
    HangarToUI getUIversion(){
        return (new HangarToUI(this));
    }
    
    private boolean spaceAvailable(){
        return (shieldBoosters.size()+weapons.size()) < maxElements;
    }
    
    public boolean addWeapon(Weapon w){
        boolean insertar=true;
        if(spaceAvailable()) weapons.add(w);
        else insertar=false;
        
        return insertar;
    }
    
    public boolean addShieldBooster(ShieldBooster s){
        boolean insertar=true;
        if(spaceAvailable()) shieldBoosters.add(s);
        else insertar=false;
        
        return insertar;
    }
    
    public ShieldBooster removeShieldBooster(int s){
        ShieldBooster shield;
        if(s>=0 && s<shieldBoosters.size())
            shield=shieldBoosters.remove(s);
        
        else shield=null;
        
        return shield;
    }
    
    public Weapon removeWeapon(int w){
        Weapon weapon;
        if(w>=0 && w<weapons.size())
            weapon=weapons.remove(w);
        
        else weapon=null;
        
        return weapon;
    }
    
    public int getMaxElements(){
        return maxElements;
    }
    
    
    public ArrayList<ShieldBooster> getShieldBoosters(){
        return shieldBoosters;
    }
    
    public ArrayList<Weapon> getWeapons(){
        return weapons;
    }
    
    
    public String toString (){
        String salida="[Hangar] -> maxElements: "+ maxElements + 
                ", shieldBoosters: "+ shieldBoosters.toString() +
                ", weapons: "+ weapons.toString();
        return salida;
    }
}

File: ruby/Hangar.rb
#encoding: utf-8
require_relative "HangarToUI"
require_relative "Weapon"
require_relative "ShieldBooster"
require_relative "WeaponType"
module Deepspace
  class Hangar
    def initialize(capacity)
      @maxElements=capacity
      @shieldBoosters=Array.new()
      @weapons=Array.new()
    end

    def self.newCopy(h)
      copy=new(h.maxElements)
      h.shieldBoosters.each do |shieldBooster|
        copy.addShieldBooster(shieldBooster)
      end
      h.weapons.each do |weapon|
        copy.addWeapon(weapon)
      end
      return copy
    end

    def getUIversion
		  return HangarToUI.new(self)
    end

    public
    def spaceAvailable
		  (@shieldBoosters.length + @weapons.length) < @maxElements
    end

    public
    def addWeapon(w)
      insertar = true;
      if(spaceAvailable)
        @weapons.push(w)
      else
        insertar = false
      end

      return insertar
      end

      def addShieldBooster(s)
        insertar = true
        if(spaceAvailable)
          @shieldBoosters.push(s)
        else
          insertar = false
        end
        return insertar
      end

    def maxElements
		  @maxElements
    end

    def shieldBoosters
		  @shieldBoosters
    end

    def weapons
		  @weapons
    end

    def removeShieldBooster(s)
		shield = nil
		if (0<=s && s<@shieldBoosters.length)
			shield = @shieldBoosters.at(s)
			@shieldBoosters.delete_at(s)
		end
		return shield
    end

    def removeWeapon(w)
		weapon = nil
		if (0<=w && w<@weapons.length) 
			weapon = @weapons.at(w)
			@weapons.delete_at(w)
		end
		return weapon
    end

    def to_s
      getUIversion.to_s
    end

  end

end
